src/ExternalSources/ExternalSourceBase.cs

```diff
@@ -0,0 +1,42 @@
+namespace Flamenco.ExternalSources;
+
+public abstract class ExternalSourceBase
+{
+    public abstract Task Download(string destinationDirectory);
+
+    public static ExternalSourceBase Create(string descriptorFilePath)
+    {
+        if (!File.Exists(descriptorFilePath))
+        {
+            throw new FileNotFoundException("Descriptor file not found.", descriptorFilePath);
+        }
+
+        var fileContentLines = File.ReadAllLines(descriptorFilePath);
+        var type = fileContentLines.FirstOrDefault(l => l.StartsWith("type"))?.Split("=").Last();
+
+        if (type is null)
+        {
+            throw new ApplicationException($"{descriptorFilePath} is missing a type.");
+        }
+
+        switch (type)
+        {
+            case "git":
+                var repositoryUrl = fileContentLines.FirstOrDefault(l => l.StartsWith("repo"))?.Split('=').Last();
+                var branch = fileContentLines.FirstOrDefault(l => l.StartsWith("branch"))?.Split('=').Last();
+                var tag = fileContentLines.FirstOrDefault(l => l.StartsWith("tag"))?.Split('=').Last();
+                var commit = fileContentLines.FirstOrDefault(l => l.StartsWith("commit"))?.Split('=').Last();
+
+                if (repositoryUrl is null)
+                {
+                    throw new ApplicationException($"Repository URL is required when descriptor file is git.");
+                }
+
+                // Reference precedence is commit, then tag, then branch. If all of them are null, the default
+                // behavior is to check out the repo's default branch.
+                return new GitExternalSource(new Uri(repositoryUrl), commit ?? (tag ?? branch));
```

Comment:
question: Why do you differentiate between commit, tag and branch? For git all three are a [commit-ish](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefcommit-ishacommit-ishalsocommittish) that git can resolve. Especially, because there is no logic that would block someone from specifying a branch-name as a tag or commit.

Reply:
For readability's sake. At the end of the day, whatever you input will end up in the same `git checkout` parameter, so there is no intrinsic distinction among the three. However, the file will make it obvious to the reader that you're specifying a branch, tag, or commit.

Reply:
I would favor using the git terminology `commitish`. This is well established, but `reference` also works and is commonly used.